src/activity.rs

```diff
@@ -11,107 +11,182 @@ use crate::frame::*;
-  pub fn from_plane<T: Pixel>(luma_plane: &Plane<T>) -> ActivityMask {
-    let PlaneConfig { width, height, .. } = luma_plane.cfg;
+  pub fn from_plane<T: Pixel>(fi: &FrameInvariants<T>, luma_plane: &Plane<T>) -> ActivityMask {
+    let granularity: usize = 3; /* Granularity of the map */
+    let act_granularity: usize = 4; /* Granularity of the activity data, if != granularity it repeats */
+    let tot_pix: usize = (1 << granularity) * (1 << granularity);
+
+    /* Aligned width and height to the activity data */
+    let width = (((luma_plane.cfg.width >> act_granularity) << act_granularity)) + (1 << act_granularity);
+    let height = (((luma_plane.cfg.height >> act_granularity) << act_granularity)) + (1 << act_granularity);
 
-    let granularity = 3;
+    let mut variances = Vec::with_capacity((width >> granularity) * (height >> granularity));
+    variances.resize((width >> granularity) * (height >> granularity), 0f64);
 
     let aligned_luma = Rect {
       x: 0_isize,
       y: 0_isize,
-      width: (width >> granularity) << granularity,
-      height: (height >> granularity) << granularity,
+      width: width,
+      height: height,
     };
     let luma = PlaneRegion::new(luma_plane, aligned_luma);
 
-    let mut variances =
-      Vec::with_capacity((height >> granularity) * (width >> granularity));
+    let mut freq_storage: Aligned<[T::Coeff; 64 * 64]> = Aligned::uninitialized();
+    let mut src_storage: Aligned<[i16; 64 * 64]> = Aligned::uninitialized();
+
+    let tx_type = TxType::DCT_DCT;
+    let tx_size = match act_granularity {
+            2 => TX_4X4,
+            3 => TX_8X8,
+            4 => TX_16X16,
+            5 => TX_32X32,
+            6 => TX_64X64,
+            _ => unreachable!(),
+        };
+
+    let mut old_avg_var = 0f64;
 
-    for y in 0..height >> granularity {
-      for x in 0..width >> granularity {
+    for y in 0..height >> act_granularity {
+      for x in 0..width >> act_granularity {
         let block_rect = Area::Rect {
-          x: (x << granularity) as isize,
-          y: (y << granularity) as isize,
-          width: 8,
-          height: 8,
+          x: (x << act_granularity) as isize,
+          y: (y << act_granularity) as isize,
+          width: 1 << act_granularity,
+          height: 1 << act_granularity,
         };
 
         let block = luma.subregion(block_rect);
 
-        let mean: f64 = block
-          .rows_iter()
-          .flatten()
-          .map(|&pix| {
-            let pix: i16 = CastFromPrimitive::cast_from(pix);
-            pix as f64
-          })
-          .sum::<f64>()
-          / 64.0_f64;
-        let variance: f64 = block
-          .rows_iter()
-          .flatten()
-          .map(|&pix| {
-            let pix: i16 = CastFromPrimitive::cast_from(pix);
-            (pix as f64 - mean).powi(2)
-          })
-          .sum::<f64>();
-        variances.push(variance);
+        let src = &mut src_storage.data[..tx_size.area()];
```

Comment:
I know you mentioned on the call you were converting this to some other sort of "hybrid"?

With this current metric, you're grabbing only the high frequency bins of a DCT. It is an interesting approximation of the contrast sensitivity metric used in PSNR-HVS, however I think it has two problems:

- The most expensive part is the forward transform. Instead of dropping the lowest coefficients, you could actually weight most of the coefficients instead.
- It is hard to evaluate this metric standalone as the patch also adds the very first usage of it.

I think it would be best to convert this to a plain pixel-domain variance and test/land the patch with that, and then upgrade it to e.g. a frequency selective DCT variance next, showing the improvements on AWCY. That would also help directly evaluate alternatives against each other, e.g. you could also implement this faster but less accurately with a hadamard transform. I think a plain variance is the best baseline to compare these options to.